phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/icons/skin/IconsListWidgetSkin.java

```diff
@@ -98,9 +98,11 @@ private void updateOldSelection(ListWidgetSelection<E> oldSelection) {
         Optional.ofNullable(oldSelection).map(ListWidgetSelection::getSelection).ifPresent(selection -> {
             final int oldValueIndex = getControl().getElements().indexOf(selection);
 
-            IconsListElement<E> oldElement = mappedElements.get(oldValueIndex);
+            if (mappedElements.size() > oldValueIndex && oldValueIndex != -1) {
```

Comment:
Why do we need this?
It shouldn't be possible for this case not to happen.

Reply:
In some cases, it happens... I don't know why

Reply:
I think it is better to locate why this happens.
Like I mentioned the "contract" for `selection` is that it should be always contained inside `getControl().getElements()`.

Reply:
The bugs appear randomly so it's hard to reproduce. It think it is because oldValueIndex is sometimes equals to -1

Reply:
The question in my opinion is, why is this the case?
I mean it shouldn't be possible.

Reply:
What if there are no selections?

Reply:
Then the `.ifPresent(selection -> {...})` in line 98 won't be called

Reply:
Probably a race condition

Reply:
I don't understand how.
This code should be executed by the JavaFX GUI thread.

Reply:
User interaction maybe

Reply:
Again I don't understand how.
All user interaction should be processed on the, according to my knowledge, single JavaFX GUI thread.
This in turn should make it impossible for race conditions to occur.

Reply:
<img width="1431" alt="capture d ecran 2018-12-30 a 21 04 53" src="https://user-images.githubusercontent.com/373645/50550784-a37c0900-0c76-11e9-93d3-b480192a3b12.png">

Reply:
<img width="1438" alt="capture d ecran 2018-12-30 a 21 06 15" src="https://user-images.githubusercontent.com/373645/50550792-cb6b6c80-0c76-11e9-90f2-a83506b14974.png">

Reply:
In fact, it happens when you select another category

Reply:
Did you set any filter settings?
Which applications did the list show?
Do you know the steps to reproduce this?

Reply:
It happens when you select another category

Reply:
Ok, the issue is, that a change to the `ListWidgetElement` list doesn't set the selection to null.
The problem here is, that setting the selection to null will still lead to the exception because the `ChangeListener` is triggered.

The only solutions seems to be to add an `if (mappedElements.size() > oldValueIndex && oldValueIndex != -1) {` to the `updateOldSelection()` method like you suggested.
Except that it needs to be added to all three list widget skins.

Still I'm unsure if there isn't a more elegant solution for the problem.

Reply:
I see. Where are the others?

Reply:
- `DetailsListWidgetSkin`
- `CompactListWidgetSkin`
- `IconsListWidgetSkin`

Reply:
Maybe also add a comment describing why the `if` is needed, otherwise it may be accidentally removed later on